stdlib/strtoull.c

```diff
@@ -30,62 +32,82 @@ uintmax_t strtoumax(const char *nptr, char **endptr, int base)
 }
 
 
-unsigned long long int strtoull(const char *nptr, char **endptr, int base)
+static unsigned long long int strtoll_common(const char *nptr, char **endptr, int base, int isUnsigned)
 {
 	const char *s;
-	unsigned long long acc;
+	unsigned long long int acc, cutoff;
 	char c;
-	unsigned long long cutoff;
 	int neg, any, cutlim;
 
 	s = nptr;
 	do {
 		c = *s++;
-	} while (isspace((unsigned char)c));
+	} while (isspace(c));
 
 	if (c == '-') {
 		neg = 1;
 		c = *s++;
-	} else {
+	}
+	else {
 		neg = 0;
-		if (c == '+')
+		if (c == '+') {
 			c = *s++;
+		}
 	}
 
 	any = 0;
-	acc = 0;
+	acc = 0ull;
 
-	if ((base == 0 || base == 16) && c == '0' && (*s == 'x' || *s == 'X')) {
+	if (((base == 0) || (base == 16)) && (c == '0') && ((s[0] == 'x') || (s[0] == 'X')) && isxdigit(s[1])) {
 		c = s[1];
 		s += 2;
 		base = 16;
-		any = 1;
 	}
 
-	if (base == 0)
-		base = c == '0' ? 8 : 10;
+	if (base == 0) {
+		base = (c == '0') ? 8 : 10;
+	}
 
+	if ((base < 2) || (base > 36)) {
+		errno = EINVAL;
+		return 0ull;
+	}
 
-	if (base < 2 || base > 36)
-		return acc;
+	if (isUnsigned != 0) {
+		cutoff = ULLONG_MAX;
+	}
+	else {
+		cutoff = (neg != 0) ? -LLONG_MIN : LLONG_MAX;
+	}
 
-	cutoff = ULONG_LONG_MAX / base;
-	cutlim = ULONG_LONG_MAX % base;
-	for ( ; ; c = *s++) {
-		if (c >= '0' && c <= '9')
+	cutlim = (int)(cutoff % base);
+	cutoff /= base;
+
+	for (;; c = *s++) {
+		if (c >= '0' && c <= '9') {
 			c -= '0';
-		else if (c >= 'A' && c <= 'Z')
+		}
+		else if (c >= 'A' && c <= 'Z') {
 			c -= 'A' - 10;
-		else if (c >= 'a' && c <= 'z')
+		}
+		else if (c >= 'a' && c <= 'z') {
 			c -= 'a' - 10;
-		else
+		}
+		else {
 			break;
```

Comment:
I checked with glibc and it doesn't set `errno` to `EINVAL` if no conversion can be performed - so perhaps it won't be very useful, but I added it.

Making code for `strtol` and `strtoll` common would be tricky, because (at least on 32-bit platforms) they use different arithmetic types for intermediate results. If we used `unsigned long long` for both of them, it would reduce performance if 32-bit conversion is requested. Although perhaps the impact wouldn't be large, because only addition and multiplication is used.